flags and research page

xithan opened the pull request:
wip

xithan committed:
working on research page

xithan committed:
add RubyMine config

xithan committed:
working on research page

xithan committed:
working on research page

ethn commented on mod/metrics/spec/set/type_plus_right/metric_value/checked_by_spec.rb:
```diff
@@ -0,0 +1,50 @@
+    end
+
+    let(:checked_by) do
+      metric_value_card.fetch trait: :checked_by
```
Are you now thinking to keep the "checked by" card, or is this an intermediate step towards  a representation that will support the double check requested status?

ethn replied:
nevermind.  I see that you're supporting that functionality on this card.

ethn commented on mod/metrics/set/type_plus_right/metric_value/value.rb:
```diff
@@ -90,6 +90,11 @@ def scored_metric?
+event :update_double_check_flag, :validate, on: [:update, :delete] do
+  return unless left.fetch trait: :checked_by
+  attach_subcard cardname.left_name.field_name(:checked_by), content: ""
+end
```
there are times when our API makes fairly nuanced functionality concise ;)

ethn commented on mod/metrics/set/type_plus_right/metric_value/checked_by.rb:
```diff
@@ -49,46 +85,76 @@ def checked_users
+    text = "Double check"
+    return text unless card.check_requested?
+    text << " #{request_icon} requested by #{card.requester}"
+    text
```
or...
```
def check_button_text
  output ["Double check", check_button_request_credit]
end

def check_button_request_credit
  return unless card.check_requested?
  " #{request_icon} requested by #{card.requester}"
end
```

ethn commented on mod/metrics/set/type_plus_right/metric_value/checked_by.rb:
```diff
@@ -1,46 +1,82 @@
-      end
-      render(@denied_view, args)
+      Auth.as_bot { card.save! }
+      render @denied_view, args
```
wait, is this where those tens of thousands of blank checked_by cards are coming from?

Do we need this?  Couldn't we get away with a few `:unknown_ok` tags on a few views instead?

ethn replied:
(I know you didn't add this.  Just sayin'.  While we're here...)

ethn replied:
It would be great if you could try to get rid of this `save!` call.  I know it's not originally your mess, but it's still adding junk data.

ethn replied:
(I think)

ethn commented on mod/metrics/view/self/research_page/source_side.haml:
```diff
@@ -0,0 +1,11 @@
+- haml_wrap do
```
you deleted a haml_wrap method above.  is this still intended here?  (I notice there's also a haml_wrap in the wagn documentation...)

xithan replied:
I moved it from wikirate it to wagn

xithan committed:
working on research pages

xithan committed:
script migration

xithan committed:
:dog2:

xithan committed:
fix success view for new answers

xithan committed:
wagn tick

xithan committed:
fix company.feature

xithan committed:
wagn tick

xithan committed:
fix research feature

xithan committed:
fix research answer feature

xithan committed:
Merge remote-tracking branch 'wagn/master' into flags

xithan committed:
wagn tick

xithan committed:
another cuke fix

xithan committed:
unknown checkbox

xithan committed:
started double check feature

xithan committed:
wagn tick

xithan committed:
fix lookup table script and import value issue

xithan committed:
workin on double check features

xithan committed:
idea settings

xithan committed:
wagn tick

xithan committed:
finished double check feature

xithan committed:
:dog2:

xithan committed:
wagn tick

xithan committed:
fix filtering

xithan committed:
remove breakpoint

xithan committed:
another breakpoint

xithan committed:
green?

xithan committed:
green??

xithan committed:
fix multi-metric research page

xithan committed:
:dog2:

xithan committed:
style hacks

ethn commented on db/migrate_cards/20170124120401_research_page.rb:
```diff
@@ -0,0 +1,31 @@
+
+    ensure_card "Unknown", codename: "unknown"
+    ensure_card "Unknown+*right+*input", content: "checkbox"
+    ensure_card "Unknown+*right+*add help",
```
Perhaps this should be `*help` and we should add this field to the edit view (in case corrections are needed)

xithan replied:
Is there an existing example for that?  A field that appears in edit view but not in edit_in_form view?

ethn commented on mod/metric_answer_lookup/set/type_plus_right/metric_value/value.rb:
```diff
@@ -1,3 +1,5 @@
+card_accessor :unknown
```
the +Unknown card is a field of the +value and not the answer?  Why is that?

Also, why implement it as a pointer rather than a boolean?

xithan replied:
I guess you mean toggle. I forgot that we have that.

xithan replied:
It's a field of value because I consider it as an interface to change the value. It's not data that we use. Actually, unknown fields are not stored in the database.  And our api doesn't support to manipulate siblings very well. It's more straight forward that +value checks if there is a +unknown subfield than to check if a +unknown is part of the same act.

xithan replied:
toggle cards have a very limited api. I have to override the editor view to get a label

ethn replied:
thanks for that explanation.  let's add `#not stored in database` as a comment on this line.

ethn commented on mod/metrics/lib/card/fixed_company_answer_query.rb:
```diff
@@ -28,7 +28,7 @@ def topic_query value
       metric_ids =
         Card.search right_plus: [Card::WikirateTopicID, { refer_to: value }],
                     return: :id
-      @restrict_to_ids[:metric_id] += metric_ids
```
I like this api change.

ethn commented on mod/metrics/set/ltype_rtype/metric/wikirate_company/record.rb:
```diff
@@ -0,0 +1,118 @@
+  def all_answers
+    @result ||=
+      Answer.fetch({ record_id: card.id }, sort_by: :year, sort_order: :desc)
+  end
```
should this be a model method?

xithan replied:
I wasn't sure if `@result` gets cleared if it is a class instance variable on `Answer`

ethn reviewed (approved):
I'm going to go ahead and merge this and deploy to staging.  But it would be great if you could review my comments. (Some of them are a week or so old, but I'm not sure you've seen them).

xithan commented:
hm, I see  comments in nine places but no one is older than an hour

xithan committed:
address github comments

xithan committed:
remove Rspec.

xithan committed:
change type of +unknown to toggle

xithan committed:
wagn tick

xithan committed:
use toggle type for +unknown

xithan committed:
Merge branch 'flags' of https://github.com/xithan/wikirate into flags

xithan committed:
wagn tick

xithan committed:
wagn tick

xithan committed:
show always unknown and check_request checkboxes

xithan committed:
Merge remote-tracking branch 'wagn/master' into flags

xithan committed:
refactor seed tasks

xithan committed:
move TypePlusRight::MetricValue::Value to Right::Value
because new cards don't have the  type plus right set

xithan committed:
remove .idea

xithan committed:
move value set back to type_plus_right

xithan committed:
Merge remote-tracking branch 'origin/flags' into flags

xithan committed:
wagn tick

xithan committed:
fixes and csv export

xithan committed:
green?

xithan committed:
green??

xithan committed:
"fill in" instead of "select"

xithan committed:
:dog2:

xithan committed:
I see specs of green :notes:
red failures too
I see them bloom
for me and you
And I think to myself
what a wondeful code
